hass-addon-sunsynk-dev/run.py

```diff
@@ -54,35 +64,85 @@ async def publish_sensors(sensors: List[Filter], *, force: bool = False) -> None
 
 async def hass_discover_sensors(serial: str, rated_power: float) -> None:
     """Discover all sensors."""
-    ents: List[Entity] = []
     dev = Device(
         identifiers=[OPT.sunsynk_id],
         name=f"Sunsynk Inverter {serial}",
         model=f"{int(rated_power/1000)}kW Inverter {serial}",
         manufacturer="Sunsynk",
     )
+    # pylint: disable=import-outside-toplevel
+    global DEVICE  # pylint: disable=global-statement
+    DEVICE = dev
+
+    ents = create_entities(SENSORS, dev)
+
+    profile_add_entities(entities=ents, device=dev)
+
+    await MQTT.connect(OPT)
+    await MQTT.publish_discovery_info(entities=ents)
+
+
+def enqueue_hass_discovery_info_update(changed_sen: Sensor, deps: List[Filter]):
+    """Add a sensor's dependants to the HASS discovery info update queue."""
+    if not DEVICE:
+        return
+
+    _LOGGER.debug(
+        "%s changed: Enqueuing discovery info updates for %s",
+        changed_sen.name,
+        ", ".join(sorted(f.sensor.name for f in deps)),
+    )
+    HASS_DISCOVERY_INFO_UPDATE_QUEUE.update((f.sensor.id, f) for f in deps)
+
+
+async def hass_update_discovery_info() -> None:
+    """Update discovery info for existing sensors"""
+    if not HASS_DISCOVERY_INFO_UPDATE_QUEUE:
+        return
+
+    ents = create_entities(HASS_DISCOVERY_INFO_UPDATE_QUEUE.values(), DEVICE)
+    HASS_DISCOVERY_INFO_UPDATE_QUEUE.clear()
+
+    await MQTT.connect(OPT)
+    await MQTT.publish_discovery_info(entities=ents, remove_entities=False)
+
+
+def create_entities(sensors: list[Filter], dev: Device) -> list[Entity]:
+    """Create HASS entities out of an existing list of filters"""
+    ents: List[Entity] = []
 
     def create_on_change_handler(filt: Filter, value_func: Callable):
         def _handler(value):
             SENSOR_WRITE_QUEUE[filt.sensor.id] = (filt, value_func(value))
 
         return _handler
 
-    for filt in SENSORS:
+    sensors = [s for s in sensors if s.sensor.id not in HIDDEN_SENSOR_IDS]
+
+    for filt in sensors:
         sensor = filt.sensor
 
         state_topic = f"{SS_TOPIC}/{OPT.sunsynk_id}/{sensor.id}"
         command_topic = f"{state_topic}_set"
 
         ent = {
             "device": dev,
-            "entity_category": "config" if isinstance(sensor, RWSensor) else "",
             "name": f"{OPT.sensor_prefix} {sensor.name}".strip(),
             "state_topic": state_topic,
             "unique_id": f"{OPT.sunsynk_id}_{sensor.id}",
             "unit_of_measurement": sensor.unit,
         }
 
+        if isinstance(sensor, RWSensor):
+            ent.update(
+                {
+                    "entity_category": "config",
+                    "icon": hass_default_rw_icon(unit=sensor.unit),
+                }
+            )
+        else:
+            ent.update({"device_class": hass_device_class(unit=sensor.unit)})
```

Comment:
Rather assign directly to the dictionary ent when its just 1/2 properties

ent["device_class"] = ...